src/tunnel.cc

```diff
@@ -1011,14 +998,15 @@ TunnelStateData::notePeerReadyToShovel(const Comm::ConnectionPointer &conn)
     retriable = false;
     server.initConnection(conn, tunnelServerClosed, "tunnelServerClosed", this);
 
-    if (!clientExpectsConnectResponse())
+    if (!clientExpectsConnectResponse)
         tunnelStartShoveling(this); // ssl-bumped connection, be quiet
     else {
         *status_ptr = Http::scOkay;
         AsyncCall::Pointer call = commCbCall(5,5, "tunnelConnectedWriteDone",
                                              CommIoCbPtrFun(tunnelConnectedWriteDone, this));
         al->reply = HttpReply::MakeConnectionEstablished();
         const auto mb = al->reply->pack();
+        // XXX: check that client connection is still open/not_closing
```

Comment:
Now that we are using a potentially stale clientExpectsConnectResponse here, should not we move this XXX up, before the whole `if` statement? AFAICT, tunnelStartShoveling(), just like this code, does not handle a closing client connection very well. It may not call client.write() directly (today) but it pretty much assumes that the client connection is alive and well.

AFAICT, the handling of that "client is closing" condition will be different in this code and inside tunnelStartShoveling() because tunnelStartShoveling() may want to finish writing early from-client data to the server, even if the client is gone. However, the XXX itself applies to both paths.

Perhaps the best thing here is not move this XXX up but _copy_ it to tunnelStartShoveling().